src/ui/components/delegation/delegateModal/Undelegate.tsx

```diff
@@ -0,0 +1,105 @@
+import type { TrackType } from '../types';
+import type { SigningAccount, VotingDelegating } from '../../../../types';
+
+import BN from 'bn.js';
+import { ChevronRightIcon, CloseIcon } from '../../../icons';
+import { Modal, Button, ButtonSecondary } from '../../../lib';
+import { useAppLifeCycle } from '../../../../lifecycle';
+import { State } from '../../../../lifecycle/types';
+import { Accounticon } from '../../accounts/Accounticon.js';
+import { SimpleAnalytics } from '../../../../analytics';
+import { useAccount } from '../../../../contexts';
+import { signAndSend } from '../../../../utils/polkadot-api';
+
+function extractBalance(state: State): BN | undefined {
```

Comment:
Could you factorize things so that there is little code duplication?
The whole modal looks pretty much duplicated. `extractBalance` could be extracted as a utility somewhere.

Reply:
The delegate and undelegate modals will evolve to be different as we add enhanced undelegate flow (by the actions from delegate page). I think makes sense to keep them separate.
I moved the extract balance to the lifecycle next to other utilities, if that makes sense.